tensordict/nn/storage.py

```diff
@@ -0,0 +1,346 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+from typing import Callable, Dict, Generic, List, Optional, TypeVar
+
+import torch
+
+import torch.nn as nn
+
+from tensordict import NestedKey, TensorDict, TensorDictBase
+from tensordict.nn.common import TensorDictModuleBase
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class TensorStorage(abc.ABC, Generic[K, V]):
+    """An Abstraction for implementing different storage.
+
+    This class is for internal use, please use derived classes instead.
+    """
+
+    def clear(self) -> None:
+        raise NotImplementedError
+
+    def __getitem__(self, item: K) -> V:
+        raise NotImplementedError
+
+    def __setitem__(self, key: K, value: V) -> None:
+        raise NotImplementedError
+
+    def __len__(self) -> int:
+        raise NotImplementedError
+
+    def contain(self, item: K) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class DynamicStorage(TensorStorage[torch.Tensor, torch.Tensor]):
+    """A Dynamic Tensor Storage.
+
+    This is a storage that save its tensors in cpu memories. It
+    expands as necessary.
+    """
+
+    def __init__(self, default_tensor: torch.Tensor):
+        self.tensor_dict: Dict[int, torch.Tensor] = {}
+        self.default_tensor = default_tensor
+
+    def clear(self) -> None:
+        self.tensor_dict.clear()
+
+    def __getitem__(self, indices: torch.Tensor) -> torch.Tensor:
```

Comment:
Good questions. 

We can put some restriction on tensors, but this means we make some assumptions on storage. I hesitate to do so, because there are some storage (soft memory like attentions) that indices can be floats. 

I left this open for future extendibility.